src/pint/simulation.py

```diff
@@ -474,36 +435,175 @@ def calculate_random_models(
             mean_vector = mean_vector[index]
         param_names = cov_matrix.get_label_names(axis=0)
 
-    f_rand = deepcopy(fitter)
-
     # scale by fac
     mean_vector = mean_vector * fac
     scaled_cov_matrix = ((cov_matrix.matrix * fac).T * fac).T
     random_models = []
-    for imodel in range(Nmodels):
+    for _ in range(Nmodels):
+        f_rand = deepcopy(fitter)
         # create a set of randomized parameters based on mean vector and covariance matrix
-        rparams_num = np.random.multivariate_normal(mean_vector, scaled_cov_matrix)
-        # scale params back to real units
-        for j in range(len(mean_vector)):
-            rparams_num[j] /= fac[j]
-        rparams = OrderedDict(zip(param_names, rparams_num))
-        f_rand.set_params(rparams)
-        phase = f_rand.model.phase(toas, abs_phase=True)
-        phases_i[imodel] = phase.int
-        phases_f[imodel] = phase.frac
-        r = pint.residuals.Residuals(toas, f_rand.model)
-        freqs[imodel] = r.get_PSR_freq(calctype="taylor")
-        if keep_models:
-            random_models.append(f_rand.model)
-            f_rand = deepcopy(fitter)
-    phases = phases_i + phases_f
-    phases0 = fitter.model.phase(toas, abs_phase=True)
-    dphase = phases - (phases0.int + phases0.frac)
-
-    if return_time:
-        dphase /= freqs
-
-    if keep_models:
-        return dphase, random_models
-    else:
-        return dphase
+        # dividing by fac brings us back to real units
+        rparams = np.random.multivariate_normal(mean_vector, scaled_cov_matrix) / fac
+        f_rand.set_params(dict(zip(param_names, rparams)))
+        random_models.append(f_rand.model)
+    return random_models
+
+
+def compute_random_model_resids_exact(fitter, models, toas, return_time=False):
+    """
+    Calculate residuals from random models exactly
```

Comment:
Docstring first lines should be on the first line, and for functions they should be full sentences. These serve as the short summary in function lists.